zerver/lib/import_realm.py

```diff
@@ -1995,3 +2008,54 @@ def add_users_to_system_user_groups(
         )
         for membership in usergroup_memberships
     )
+
+
+def check_migration_status(migration_status_file: str) -> None:
+    # Each server might have a different set of apps installed (e.g., Zulip Cloud has
+    # zilencer/corporate, but self-hosted might not). This function only checks for
+    # migration mismatches in apps installed on both servers.
+    local_migrations = parse_migration_status(get_migration_status())
+    exported_migrations = parse_migration_status(migration_status_file)
+
+    mismatched_migrations_log = [
+        f"Migrations for the '{app}' app doesn't match; showing diff.\n"
+        + diff_strings(local_migrations[app], exported_migrations[app])
+        for app in exported_migrations
+        if local_migrations.get(app) and local_migrations[app] != exported_migrations[app]
+    ]
+
+    match = re.search(r"ZULIP_VERSION=([^\s]+)", migration_status_file)
+    exported_zulip_version = match.group(1) if match else "Unknown"
+
+    if mismatched_migrations_log:
+        raise AssertionError(
+            f"""
+---
+Mismatch detected in applied migrations between the exported file and this server.
+Ensure both servers are running the same Zulip version!
+
+Zulip versions:
+* Exported realm: {exported_zulip_version}
+* This server: {ZULIP_VERSION}
+---
+"""
+            + "\n".join(mismatched_migrations_log)
+        )
```

Comment:
Can you post sample output from this failure?

Reply:
OK, took a look. I think we probably want to just print lines for the migrations that differ; having 2 screenfuls of output where the only difference is 3 added migrations seems like a bad experience compared with just printing those differences.

Reply:
I ended up opting for raising `SystemExit` instead of `AssertionError` for this case, it doesn't output any traceback and only the diffs error message. Is there any technical concern that I should be aware of when using `SystemExit`?

The error message itself should be more concise now with only 1 context line. The updated error logs are in the PR description and the error messages themselves are in these fixtures:
[missing_migrations_error.txt](https://github.com/zulip/zulip/pull/31681/files#diff-396667a03beace09e335749f3cf46892323e1a5442ec4836e64c38cb09440168) & [unapplied_migrations_error.txt](https://github.com/zulip/zulip/pull/31681/files#diff-a27ed2d8a17941e48d5b4e05c87a29f31d6270a8d0bac1e7b655a829bdd2aeae)